src/ring_theory/polynomial/eisenstein.lean

```diff
@@ -0,0 +1,187 @@
+/-
+Copyright (c) 2022 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import ring_theory.eisenstein_criterion
+
+/-!
+# Eisenstein polynomials
+Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R` is
+*Eisenstein at `𝓟`* if `f.leading_coeff ∉ 𝓟`, `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟` and
+`f.coeff 0 ∉ 𝓟 ^ 2`. In this file we gather miscellaneous results about Eisenstein polynomials.
+
+## Main definitions
+* `polynomial.is_eisenstein_at f 𝓟` : the property of being Eisenstein at `𝓟`.
+
+## Main results
+* `polynomial.is_eisenstein_at.irreducible` : if a primitive `f` satisfies `f.is_eisenstein_at 𝓟`,
+  where `𝓟.is_prime`, then `f` is irreducible.
+
+## Implementation details
+We also define a notion `is_weakly_eisenstein_at` requiring only that
+`∀ n < f.nat_degree → f.coeff n ∈ 𝓟`. This makes certain results slightly more general and it is
+useful since it is sometimes better behaved (for example it is stable under `polynomial.map`).
+
+-/
+
+universes u v
+
+variables {R : Type u}
+
+open ideal algebra finset
+
+open_locale big_operators
+
+namespace polynomial
+
+/-- Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R`
+is *weakly Eisenstein at `𝓟`* if `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟`. -/
+@[mk_iff] structure is_weakly_eisenstein_at [comm_semiring R] (f : polynomial R) (𝓟 : ideal R) :
+  Prop := (mem : ∀ {n}, n < f.nat_degree → f.coeff n ∈ 𝓟)
+
+/-- Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R`
+is *Eisenstein at `𝓟`* if `f.leading_coeff ∉ 𝓟`, `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟` and
+`f.coeff 0 ∉ 𝓟 ^ 2`. -/
+@[mk_iff] structure is_eisenstein_at [comm_semiring R] (f : polynomial R) (𝓟 : ideal R) : Prop :=
+(leading : f.leading_coeff ∉ 𝓟)
+(mem : ∀ {n}, n < f.nat_degree → f.coeff n ∈ 𝓟)
```

Comment:
You could also `extends is_weakly_eisenstein_at`, which would give you the projection for free.

Reply:
This was my first idea, but then `@[mk_iff]` will produce an ugly lemma (saying that it is Eisenstein iff it is weakly Eistenstein plus...), right?

Reply:
Aah, that could well be. It makes sense to keep it like this then.